keps/sig-api-machinery/4355-coordinated-leader-election/README.md

```diff
@@ -0,0 +1,1107 @@
+
+### Troubleshooting
+
+<!--
+This section must be completed when targeting beta to a release.
+
+For GA, this section is required: approvers should be able to confirm the
+previous answers based on experience in the field.
+
+The Troubleshooting section currently serves the `Playbook` role. We may consider
+splitting it into a dedicated `Playbook` document (potentially with some monitoring
+details). For now, we leave it here.
+-->
+
+###### How does this feature react if the API server and/or etcd is unavailable?
+
+###### What are other known failure modes?
+
+<!--
+For each of them, fill in the following information by copying the below template:
+  - [Failure mode brief description]
+    - Detection: How can it be detected via metrics? Stated another way:
+      how can an operator troubleshoot without logging into a master or worker node?
+    - Mitigations: What can be done to stop the bleeding, especially for already
+      running user workloads?
+    - Diagnostics: What are the useful log messages and their required logging
+      levels that could help debug the issue?
+      Not required until feature graduated to beta.
+    - Testing: Are there any tests for failure mode? If not, describe why.
+-->
+
+###### What steps should be taken if SLOs are not being met to determine the problem?
+
+## Implementation History
+
+<!--
+Major milestones in the lifecycle of a KEP should be tracked in this section.
+Major milestones might include:
+- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
+- the `Proposal` section being merged, signaling agreement on a proposed design
+- the date implementation started
+- the first Kubernetes release where an initial version of the KEP was available
+- the version of Kubernetes where the KEP graduated to general availability
+- when the KEP was retired or superseded
+-->
+
+## Drawbacks
+
+<!--
+Why should this KEP _not_ be implemented?
+-->
+
+## Alternatives
+
+When evaluating alternatives, note that if we decide in the future to improve
+the algorithm, fix a bug in the algorithm, or change the criteria for how
+leaders are elected, our decision on where to put the code has a huge impact our
+how the change is rolled out.
+
+For example, it will be much easier change in a controller in the kube-apiserver
+than in client-go library code distributed to elected controllers, because once
+it is distributed into controllers, especially 3rd party controllers, any change
+requires updating client-go and then updating all controllers to that version of
+client-go.
+
+### Component instances pick a leader without a coordinator
+
+- A candidates is picked at random to be an election coordinator, and the
+   coordinator picks the leader:
+  - Components race to claim the lease
+  - If a component claims the lease, the first thing it does is check the
+    identity leases to see if there is a better leader
+  - If it finds a better lease, it assigns the lease to that component instead
```

Comment:
what happens when a better leader appears later. Is there preemption?

Reply:
Yes.  The preemption is collaborative/non-byzantine.  The way it works is that the election controller writes a "end of term" marker on the lease.  When the current leader sees this, it stops renewing the lease and it expires.  Once the lease expires,  a new election can start.  "end of term"  is only considered valid on a non-expired lease, so once a lease expires, the next time the lease is claimed, "end of term" can be cleared.

There is some lag in this approach, but it's easy to reason about the correctness...